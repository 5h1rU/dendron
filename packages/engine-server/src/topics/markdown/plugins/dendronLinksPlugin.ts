import { DNoteLoc } from "@dendronhq/common-all";
import { removeMDExtension } from "@dendronhq/common-server";
import _ from "lodash";
import RemarkParser, { Eat } from "remark-parse";
import { Parser, Processor } from "unified";
import { Node } from "unist";
import { LinkOpts, WikiLinkData, WikiLinkProps } from "./types";

export const LINK_REGEX = /^\[\[(.+?)\]\]/;

declare class RemarkParserClass implements Parser {
  Parser: RemarkParserClass;
  prototype: RemarkParserClass;
  parse(): Node;
  blockMethods: string[];
  inlineTokenizers: {
    [key: string]: RemarkParser.Tokenizer;
  };

  inlineMethods: string[];
}

function locator(value: string, fromIndex: number) {
  return value.indexOf("[", fromIndex);
}

type PluginOpts = {
  permalinks?: string[];
  pageResolver?: (name: string) => string[];
  newClassName?: string;
  wikiLinkClassName?: string;
  hrefTemplate?: (permalink: string) => string;
  aliasDivider?: string;
  // dendron opts
  replaceLink?: { from: DNoteLoc; to: DNoteLoc };
  convertObsidianLinks?: boolean;
} & LinkOpts;

export { PluginOpts as DendronLinksOpts };

function genLinkValue(value: string, data: WikiLinkData) {
  if (data.anchorHeader) {
    value = value += `#${data.anchorHeader}`;
  }
  return value;
}

/**
 * Matches wiki-links
 * @param opts
 */
export function dendronLinksPlugin(opts: Partial<PluginOpts> = {}) {
  const permalinks = opts.permalinks || [];
  const defaultPageResolver = (name: string) => [
    name.replace(/ /g, "").toLowerCase(),
  ];
  const pageResolver = opts.pageResolver || defaultPageResolver;
  const newClassName = opts.newClassName || "new";
  const wikiLinkClassName = opts.wikiLinkClassName || "internal";
  const defaultHrefTemplate = (permalink: string) => `#/page/${permalink}`;
  const hrefTemplate = opts.hrefTemplate || defaultHrefTemplate;
  const aliasDivider = opts.aliasDivider || "|";
  const copts = _.defaults(opts, { convertObsidianLinks: false });

  function isAlias(pageTitle: string) {
    return pageTitle.indexOf(aliasDivider) !== -1;
  }

  function parseAliasLink(pageTitle: string) {
    const [alias, value] = pageTitle.split(aliasDivider);
    return { alias, value };
  }
  function parseAnchorIfExist(link: string) {
    if (link.indexOf("#") !== -1) {
      return link.split("#");
    } else {
      return [link, undefined];
    }
  }

  function parsePageTitle(pageTitle: string) {
    let out: WikiLinkProps = {
      value: pageTitle,
      alias: pageTitle,
    };
    if (isAlias(pageTitle)) {
      out = parseAliasLink(pageTitle);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#");
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!isAlias(pageTitle)) {
        out.alias = value;
      }
    }
    return out;
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);

    if (match) {
      // contents of the link
      const pageName = match[1].trim();
      const { value, alias, anchorHeader } = parsePageTitle(pageName);

      const pagePermalinks = pageResolver(value);
      let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
      const exists = permalink !== undefined;

      if (!exists) {
        permalink = pagePermalinks[0];
      }

      let classNames = wikiLinkClassName;
      if (!exists) {
        classNames += " " + newClassName;
      }
      // normalize
      if (permalink) {
        permalink = removeMDExtension(permalink);
      }

      return eat(match[0])({
        type: "wikiLink",
        value,
        data: {
          alias,
          toMdEnhancedPreview: opts.toMdEnhancedPreview,
          permalink,
          anchorHeader,
          exists,
          hName: "a",
          hProperties: {
            className: classNames,
            href: hrefTemplate(permalink as string),
          },
          hChildren: [
            {
              type: "text",
              value: alias,
            },
          ],
        } as WikiLinkData,
      });
    }
    return;
  }

  inlineTokenizer.locator = locator;

  // @ts-ignore
  let _this = (this as any) as RemarkParserClass;
  const Parser: RemarkParser.Parser = _this.Parser;
  // @ts-ignore
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  // @ts-ignore
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.wikiLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "wikiLink");

  // Stringify for wiki link
  // @ts-ignore
  const Compiler: Processor["Compiler"] = _this.Compiler;

  if (Compiler != null) {
    const visitors = Compiler.prototype.visitors;
    if (visitors) {
      visitors.wikiLink = function (node: Node) {
        debugger;
        const data = node.data as WikiLinkData;
        if (!node || !node.data || !node.data.alias) {
          throw Error(`no alias found: ${JSON.stringify(node)}`);
        }
        if (data.useId) {
          if (!data.note) {
            throw Error("no note");
          }
          node.value = data.note.id;
        }
        if (opts.replaceLink && opts.replaceLink.from.fname === node.value) {
          // TODO: check for case
          node.value = opts.replaceLink.to.fname;
          if (node.data.alias === opts.replaceLink.from.fname) {
            node.data.alias = opts.replaceLink.to.fname;
          }
        }
        let nodeValue = genLinkValue(node.value as string, data);
        if (data.toMd) {
          return `[${data.alias}](${data.prefix || ""}${nodeValue})`;
        }
        if (data.forNoteRefInPreview) {
          return `${data.prefix || ""}${node.value}.md`;
        }
        if (data.forNoteRefInSite) {
          return `${data.prefix || ""}${node.value}.html`;
        }
        if (data.toHTML) {
          return `${data.prefix || ""}${node.value}.html`;
        }

        if (node.data.alias !== node.value) {
          return `[[${node.data.alias}${aliasDivider}${nodeValue}]]`;
        }
        if (copts.convertObsidianLinks) {
          nodeValue = _.replace(nodeValue as string, /\//g, ".");
        }
        if (data.toMdEnhancedPreview) {
          return `[[${parseAnchorIfExist(node.value as string)[0]}]]`;
        }
        return `[[${nodeValue}]]`;
      };
    }
  }
}
