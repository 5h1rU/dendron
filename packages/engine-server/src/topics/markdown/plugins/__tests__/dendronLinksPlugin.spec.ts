import unified from "unified";
import { dendronLinksPlugin } from "../dendronLinksPlugin";
import markdownParse from "remark-parse";

function getProcessor() {
  let processor: unified.Processor | null = null;
  return (processor = unified()
    .use(markdownParse, { gfm: true })
    .use(dendronLinksPlugin));
}

describe("basic", () => {
  test("init", () => {
    const resp = getProcessor().parse(`[["foo.md"]]`);
    expect(resp).toMatchSnapshot();
    // @ts-ignore
    expect(resp.children[0].children[0].type).toEqual("wikiLink");
  });

  test("doesn't parse inline code block", () => {
    const resp = getProcessor().parse("`[[foo.md]]`");
    expect(resp).toMatchSnapshot("bond");
    // child1 paragraph, child2 link
    // @ts-ignore
    // expect(resp.children[0].children[0].data.link).not.toEqual({
    //   type: "file",
    //   name: "foo",
    //   anchorStart: undefined,
    //   anchorEnd: undefined,
    // });
  });
});
