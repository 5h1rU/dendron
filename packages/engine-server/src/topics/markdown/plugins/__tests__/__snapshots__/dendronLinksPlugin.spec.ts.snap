// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic doesn't parse inline code block: bond 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "inlineCode",
          "value": "[[foo.md]]",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 13,
          "line": 1,
          "offset": 12,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic init 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "\\"foo.md\\"",
            "anchorHeader": undefined,
            "exists": false,
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "\\"foo.md\\"",
              },
            ],
            "hName": "a",
            "hProperties": Object {
              "className": "internal new",
              "href": "#/page/\\"foo",
            },
            "permalink": "\\"foo",
          },
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "wikiLink",
          "value": "\\"foo.md\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 13,
          "line": 1,
          "offset": 12,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
